### DISCUSSION TITLE
Implement Deserialize trait on Chart (and more)?

### DISCUSSION THREAD

**Discussion by jasoncg** [yuankunzhang/charming#128] [Ideas]
I've just started trying out Charming and it seems it'd be a good fit for my project. One thing though is I'm looking to serialize and deserialize charts, but the Chart struct doesn't implement the `Deserialize` trait, only `Serialize`. I'm doing some calculations in a web worker, which I pass back to an iframe over a MessagePort. I built the Chart object in the worker, serialized it with serde-wasm-bindgen, posted it to the iframe, and then ran into the problem that I couldn't deserialize it with serde-wasm-bindgen.

I considered three options:

1. Work with the serialized Chart JsValue (serde-wasm-bindgen), bypassing Charming and sending the object directly to Echarts. I did experiment a bit with making Charming's Echarts::set_option function public to allow me to send the JsValue directly and this worked, but that's a rather hacky approach.

2. Send my raw data from the web worker to the iframe, and build the Chart object there. At least this wouldn't require any changes to Charming or hacks, but it's a bit convoluted with regard to my application. I'd have to replicate several structures that already exist in Charming to capture the configuration of the chart. 

3. **Add the Deserialize trait to Chart / throughout Charming**. I tried this and it seems to work, but I haven't made any complex charts yet. I did create a couple simple tests though in [tests/deserialize.rs
](https://github.com/jasoncg/charming/blob/deserialize-chart/charming/tests/deserialize.rs)

Here's my branch with all the changes:
https://github.com/jasoncg/charming/tree/deserialize-chart

For the most part it was just adding Deserialize, but some structures required custom deserializers.

Data structures where I had to implement a custom Deserialize trait include:
- datatype::Dataset
- series::Series
- series::ThemeRiverData
- element::axis_line
- element::BoundaryGap
- element::FontFamily
- element::Icon
- element::MarkLineVariant
- element::Padding
- element::RawString
- element::Symbol

The other issue I ran in to was some vectors have `skip_serializing_if = "Vec::is_empty"`, which was causing problem with deserialization. So for all of those I set `#[serde(default)]` so that they can deserialize with empty vectors.

Is there another mechanism for deserializing that I'm missing? Should I submit a PR to merge these changes?

**Comment by LukaOber** [comment]
Having the ability to deserialize would be great. I would be very happy to take a look at a PR implementing these changes. I am currently very busy with work, so please excuse me if I won't have the time to review the changes in the upcoming days.

**Reply by jasoncg** [reply]
Understood, I submitted a PR #129. Thanks!

**Comment by mmemcormier** [comment]
@LukaOber  are you still interested in getting deserialization implemented? In a sense it is a commitment that all future types will need to implement deserialization as well. I can submit a PR with tests with if this is desired.

**Reply by LukaOber** [reply]
I think we should definitely implement deserialize as some users of the library need it. If you are able to open up a PR I would greatly appreciate it and would provide help when you need it.

**Comment by LukaOber** [comment]
Deserialize was implemented in #172
